File: quant1x/config/strategy-parameter.go
package config

// StrategyParameter mirrors key fields from C++ StrategyParameter
type StrategyParameter struct {
	Id                          uint64         `yaml:"id" json:"id"`
	Auto                        bool           `yaml:"auto" json:"auto"`
	Name                        string         `yaml:"name" json:"name"`
	Flag                        string         `yaml:"flag" json:"flag"`
	Session                     TradingSession `yaml:"time" json:"time"`
	Weight                      float64        `yaml:"weight" json:"weight"`
	Total                       int            `yaml:"total" json:"total"`
	PriceCageRatio              float64        `yaml:"price_cage_ratio" json:"price_cage_ratio"`
	MinimumPriceFluctuationUnit float64        `yaml:"minimum_price_fluctuation_unit" json:"minimum_price_fluctuation_unit"`
	FixedSlippageForSell        float64        `yaml:"fixed_slippage_for_sell" json:"fixed_slippage_for_sell"`
	FeeMax                      float64        `yaml:"fee_max" json:"fee_max"`
	FeeMin                      float64        `yaml:"fee_min" json:"fee_min"`
	Sectors                     []string       `yaml:"sectors" json:"sectors"`
	IgnoreMarginTrading         bool           `yaml:"ignore_margin_trading" json:"ignore_margin_trading"`
	HoldingPeriod               int            `yaml:"holding_period" json:"holding_period"`
	SellStrategy                uint64         `yaml:"sell_strategy" json:"sell_strategy"`
	FixedYield                  float64        `yaml:"fixed_yield" json:"fixed_yield"`
	TakeProfitRatio             float64        `yaml:"take_profit_ratio" json:"take_profit_ratio"`
	StopLossRatio               float64        `yaml:"stop_loss_ratio" json:"stop_loss_ratio"`
	LowOpeningAmplitude         float64        `yaml:"low_opening_amplitude" json:"low_opening_amplitude"`
	HighOpeningAmplitude        float64        `yaml:"high_opening_amplitude" json:"high_opening_amplitude"`
	Rules                       RuleParameter  `yaml:"rules" json:"rules"`
	ExcludeCodes                []string       `yaml:"exclude_codes" json:"exclude_codes"`
}

// defaultStrategyParameter returns StrategyParameter with C++ defaults
func defaultStrategyParameter() StrategyParameter {
	return StrategyParameter{
		Id:                          1,
		Auto:                        false,
		Name:                        "",
		Flag:                        "",
		Weight:                      0.0,
		Total:                       3,
		PriceCageRatio:              0.00,
		MinimumPriceFluctuationUnit: 0.00,
		FixedSlippageForSell:        FixedSlippageForSell,
		FeeMax:                      20000.00,
		FeeMin:                      10000.00,
		Sectors:                     nil,
		IgnoreMarginTrading:         true,
		HoldingPeriod:               1,
		SellStrategy:                117,
		FixedYield:                  0.0,
		TakeProfitRatio:             15.0,
		StopLossRatio:               -2.0,
		LowOpeningAmplitude:         0.618,
		HighOpeningAmplitude:        0.382,
		Rules:                       defaultRuleParameter(),
		ExcludeCodes:                nil,
	}
}

File: quant1x/config/strategy_parameter.cpp
#include <quant1x/config/strategy_parameter.h>
#include <quant1x/data/market/instruments.h>
#include <quant1x/data/exchange/margin_trading.h>

namespace config {

    // 初始化排除列表
    void StrategyParameter::initExclude() {
        if (!excludeCodes.empty()) {
            return;
        }

        std::vector<std::string> tempExcludeCodes;
        for (const auto& v : Sectors) {
            std::string sectorCode = strings::trim(v);
            if (sectorCode.starts_with(sectorIgnorePrefix)) {
                sectorCode = strings::trim(sectorCode.substr(sectorPrefixLength));
                auto blockInfo = exchange::get_sector_info(sectorCode);
                if (blockInfo.has_value()) {
                    tempExcludeCodes.insert(tempExcludeCodes.end(),
                                            blockInfo->ConstituentStocks.begin(),
                                            blockInfo->ConstituentStocks.end());
                }
            }
        }

        // 去重
        std::unordered_set<std::string> uniqueCodes(tempExcludeCodes.begin(), tempExcludeCodes.end());
        excludeCodes.assign(uniqueCodes.begin(), uniqueCodes.end());
    }

    // 过滤股票代码
    std::vector<std::string> StrategyParameter::Filter(const std::vector<std::string>& codes) {
        initExclude();

        // 过滤需要忽略的板块成份股
        std::vector<std::string> newCodeList;
        std::copy_if(codes.begin(), codes.end(), std::back_inserter(newCodeList),
                     [this](const std::string& s) {
                         return std::find(excludeCodes.begin(), excludeCodes.end(), s) == excludeCodes.end();
                     });

        // 去重
        std::unordered_set<std::string> uniqueCodes(newCodeList.begin(), newCodeList.end());
        newCodeList.assign(uniqueCodes.begin(), uniqueCodes.end());

        // 排序
        std::sort(newCodeList.begin(), newCodeList.end());

        if (IgnoreMarginTrading) {
            // 过滤两融
            auto marginTradingList = exchange::MarginTradingList();
            std::unordered_set<std::string> marginSet(marginTradingList.begin(), marginTradingList.end());

            std::vector<std::string> filteredList;
            std::copy_if(newCodeList.begin(), newCodeList.end(), std::back_inserter(filteredList),
                         [&marginSet](const std::string& s) {
                             return marginSet.find(s) == marginSet.end();
                         });
            return filteredList;
        }
        return newCodeList;
    }

    // 取得可以交易的证券代码列表
    std::vector<std::string> StrategyParameter::StockList() {
        std::vector<std::string> codes;
        for (const auto& v : Sectors) {
            std::string sectorCode = strings::trim(v);
            if (!sectorCode.starts_with(sectorIgnorePrefix)) {
                auto blockInfo = exchange::get_sector_info(sectorCode);
                if (blockInfo.has_value()) {
                    codes.insert(codes.end(),
                                 blockInfo->ConstituentStocks.begin(),
                                 blockInfo->ConstituentStocks.end());
                }
            }
        }

        if (codes.empty()) {
            codes = instruments::GetStockCodeList();
        }
        return Filter(codes);
    }

    std::ostream &operator<<(std::ostream &os, const StrategyParameter &parameter) {
        os << "{Id: " << parameter.Id  << "\n"
           << " Auto: " << parameter.Auto  << "\n"
           << " Name: " << parameter.Name << "\n"
           << " Flag: " << parameter.Flag << "\n"
           << " Session: " << parameter.Session.ToString() << "\n"
           << " Weight: " << parameter.Weight << "\n"
           << " Total: " << parameter.Total << "\n"
           << " PriceCageRatio: " << parameter.PriceCageRatio << "\n"
           << " MinimumPriceFluctuationUnit: " << parameter.MinimumPriceFluctuationUnit << "\n"
           << " FeeMax: " << parameter.FeeMax << "\n"
           << " FeeMin: " << parameter.FeeMin << "\n"
           //<< " Sectors: " << parameter.Sectors << "\n"
           << " IgnoreMarginTrading: " << parameter.IgnoreMarginTrading << "\n"
           << " HoldingPeriod: " << parameter.HoldingPeriod << "\n"
           << " SellStrategy: " << parameter.SellStrategy << "\n"
           << " FixedYield: " << parameter.FixedYield << "\n"
           << " TakeProfitRatio: " << parameter.TakeProfitRatio << "\n"
           << " StopLossRatio: " << parameter.StopLossRatio << "\n"
           << " LowOpeningAmplitude: " << parameter.LowOpeningAmplitude << "\n"
           << " HighOpeningAmplitude: " << parameter.HighOpeningAmplitude << "\n"
           << " Rules: " << parameter.Rules
           //<< " excludeCodes: " << parameter.excludeCodes;
           << "}";
        return os;
    }
}